shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java

```diff
@@ -36,20 +34,33 @@ public InputConnectionAdaptor(
         View view,
         int client,
         TextInputChannel textInputChannel,
-        Editable editable
+        Editable editable,
+        Runnable onConnectionClosed
     ) {
         super(view, true);
         mFlutterView = view;
         mClient = client;
         this.textInputChannel = textInputChannel;
         mEditable = editable;
+        this.onConnectionClosed = onConnectionClosed;
         mBatchCount = 0;
         // We create a dummy Layout with max width so that the selection
         // shifting acts as if all text were in one line.
         mLayout = new DynamicLayout(mEditable, new TextPaint(), Integer.MAX_VALUE, Layout.Alignment.ALIGN_NORMAL, 1.0f, 0.0f, false);
         mImm = (InputMethodManager) view.getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
     }
 
+    @Override
+    public void closeConnection() {
+        super.closeConnection();
+
+        textInputChannel.onConnectionClosed(mClient);
+
+        if (onConnectionClosed != null) {
```

Comment:
nit: This is final and never null, so I think it's okay to just always run this. Not a big deal and it doesn't hurt to be safe though.

related: annotations may be worthwhile in the constructor?

Reply:
Good point, I'll add annotations.

Reply:
The parameters are now annotated and the if-statement has been removed.